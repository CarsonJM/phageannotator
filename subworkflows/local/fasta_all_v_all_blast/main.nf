//
// Compare sequences by performing an all-v-all BLAST
//
include { BLAST_MAKEBLASTDB         } from '../../../modules/nf-core/blast/makeblastdb/main'
include { BLAST_BLASTN              } from '../../../modules/nf-core/blast/blastn/main'
include { CHECKVANICLUSTER_ANICALC  } from '../../../modules/local/checkvanicluster/anicalc/main'

workflow FASTA_ALL_V_ALL_BLAST {

    take:
    fasta_gz    // [ [ meta ], fasta.gz ]   , assemblies/genomes (mandatory)

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Make BLASTN database
    //
    ch_blast_db = BLAST_MAKEBLASTDB ( fasta_gz ).db
    ch_versions = ch_versions.mix( BLAST_MAKEBLASTDB.out.versions )

    //
    // MODULE: Perform BLASTN all-v-all alignment
    //
    ch_blast_txt = BLAST_BLASTN ( fasta_gz , ch_blast_db ).txt
    ch_versions = ch_versions = ch_versions.mix( BLAST_MAKEBLASTDB.out.versions )

    //
    // MODULE: Calculate average nucleotide identity (ANI) and alignment fraction (AF) based on BLAST
    //
    ch_blast_ani_tsv    = CHECKVANICLUSTER_ANICALC ( ch_blast_txt ).ani
    ch_versions         = ch_versions.mix( CHECKVANICLUSTER_ANICALC.out.versions )

    emit:
    blast_ani_tsv   = ch_blast_ani_tsv  // [ [ meta ], ani.tsv ]   , TSV file containing BLAST all-v-all ANI and AF values
    ch_versions     = ch_versions       // [ versions.yml ]

}
