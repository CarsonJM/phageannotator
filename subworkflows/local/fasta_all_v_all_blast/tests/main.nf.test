nextflow_workflow {

    name "Test Subworkflow: FASTA_ALL_V_ALL_BLAST"
    script "subworkflows/local/fasta_all_v_all_blast/main.nf"
    workflow "FASTA_ALL_V_ALL_BLAST"


    test("fasta.gz") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file("$projectDir/assets/test-datasets/fastas/MEGAHIT-BUFF_01_B.contigs.fa.gz", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
