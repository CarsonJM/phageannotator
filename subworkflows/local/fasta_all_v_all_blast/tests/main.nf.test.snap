{
    "fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_ani.tsv:md5,ee66197f71cf24a9ea45a007e8b36216"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3a67b197cc055b730a0f1aa4d0cd31c0",
                    "versions.yml:md5,3a67b197cc055b730a0f1aa4d0cd31c0",
                    "versions.yml:md5,6eb3d40e5dcf28a68a47eb11190e16bd"
                ],
                "blast_ani_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test_ani.tsv:md5,ee66197f71cf24a9ea45a007e8b36216"
                    ]
                ],
                "ch_versions": [
                    "versions.yml:md5,3a67b197cc055b730a0f1aa4d0cd31c0",
                    "versions.yml:md5,3a67b197cc055b730a0f1aa4d0cd31c0",
                    "versions.yml:md5,6eb3d40e5dcf28a68a47eb11190e16bd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-02T10:19:48.024860225"
    }
}